### DISCUSSION TITLE
Custom hill shading algorithm

### DISCUSSION THREAD

**Discussion by Sublimis** [mapsforge/mapsforge#1512] [Q&A] [answered]
Creating a custom hill shading algorithm is currently severely restricted due to HgtCache.HgtFileInfo class being package private. Would it be possible to somehow lift this restriction (making the class public would suffice, but it might not be the most elegant solution)?

**Comment by devemux86** [comment]
@Sublimis

What customizations do you have in mind? Hillshade colors or something else?
Add more parameters in the algorithms?

It would be nice if we integrate these customizations so all users can use them.
Making some classes public is also possible.

**Comment by Sublimis** [comment]
So, we just started experimenting on whether hill shading can be improved (not sure about the outcome), and wanted to gain full control over the values written to the byte buffer used for creating `AndroidHillshadingBitmap` (ALPHA_8 config). However, this proved to be impossible because subclassing either `public abstract class AbsShadingAlgorithmDefaults implements ShadingAlgorithm` or even the `public interface ShadingAlgorithm` is futile.

To be more precise, implementing/overriding `protected abstract byte[] convert(ByteBuffer map, int axisLength, int rowLen, int padding, HgtCache.HgtFileInfo source)` -or- `public RawShadingResult transformToByteBuffer(HgtCache.HgtFileInfo source, int padding)` is impossible due to the aforementioned issue with HgtCache.HgtFileInfo privacy which is taken in as a parameter...

Many thanks!

**Reply by devemux86** [reply]
@Sublimis 

If you start experimenting on how to improve the hillshading,
it's better to clone Mapsforge, do tests and when come up with a final result that is better than the current implementation,
we can discuss changes and improvements that can be merged into the library?

**Reply by Sublimis** [reply]
Okay, will let you know.

(Btw. just noticed, Jitpack build for the latest commit failed)

**Reply by devemux86** [reply]
> (Btw. just noticed, Jitpack build for the latest commit failed)

The first time JitPack builds the artifacts, if they have not been previously requested.
So it may take some time, now it seems to work correctly.

**Reply by Sublimis** [reply]
Please see here: [https://jitpack.io/com/github/mapsforge/mapsforge/204c94093a/build.log](https://jitpack.io/com/github/mapsforge/mapsforge/204c94093a/build.log) 🤷

**Reply by devemux86** [reply]
JitPack master-SNAPSHOT was built green and also tested in [vtm-android-example](https://github.com/mapsforge/vtm/tree/master/vtm-android-example):

https://jitpack.io/#mapsforge/mapsforge/master-SNAPSHOT

**Reply by demiantres** [reply]
Is the new hillshading implementation ready (except bug fixes), or are the bigger changes expected?

**Reply by devemux86** [reply]
@demiantres 

All hillshading improvements and the new algorithm are already available in the `master` branch.

Some extra improvements are prepared by @Sublimis, see the discussion here:
- https://github.com/mapsforge/mapsforge/pull/1521#issuecomment-2332217516

**Comment by Sublimis** [comment]
Here are the results.
All screenshots were made using the latest Motorider theme included with the library (except for the Google screenshot, of course).

The next several screenshots are using the hill shading magnitude scales of both 3 and 1 for better comparison.

All screenshots are for illustrative purposes.

First our hill shading algorithm as finally used in the app (scale = 3):

![20240815_ours_scale=3](https://github.com/user-attachments/assets/6bd9aba5-19e3-45c9-8735-c9a4b902e9e7)


Compare with the default Simple algorithm (scale = 3),

![20240815_simple_default_scale=3](https://github.com/user-attachments/assets/3017393b-a1c5-4dfd-8e6e-f96d3f854529)


the default default Simple algorithm (scale = 1),

![20240815_simple_default_scale=1](https://github.com/user-attachments/assets/ff57463e-08a6-4c9b-8ecf-e2c84e1947e1)


the default Diffuse Light algorithm (scale = 3) -- please observe how particularly bad the Lim and Raša fjords look like:

![20240815_diffuse_default_scale=3](https://github.com/user-attachments/assets/424c380f-31ed-4f3c-a62d-628f3230be8a)


and the default default Diffuse Light algorithm (scale = 1):

![20240815_diffuse_default_scale=1](https://github.com/user-attachments/assets/b831ac0f-b035-4fc1-98cf-15b8feba28c2)


Here is our algorithm with scale = 1 (just for completeness):

![20240815_ours_scale=1](https://github.com/user-attachments/assets/8f96d5d1-b132-4d07-9f9d-bd82b471a896)


For completeness, here is the screenshot of Google terrain map ("legacy" renderer):

![20240815_google_terrain](https://github.com/user-attachments/assets/3c4cb25d-2cf4-4aaf-ad70-6fe72ff3dada)


and again the Motorider map but without any hill shading:

![20240815_none](https://github.com/user-attachments/assets/c8fd778a-6ce8-48d6-b544-44f005b54f37)



Our biggest objection with the current hill shading algorithms is that they always show a gray overlay over an otherwise flat terrain, reducing contrast of the map, making it darker and, well, uglier.

Our algorithm is based on the Simple one, nothing too fancy, but because of a change in paradigm it won't ever put a shade over a flat terrain. Additional bonus is that this also makes hills and fjords look nicer, and closer to what we see in Google maps (we can provide additional screenshots). If you're interested in including the code in the library, we need a few more days to pretty up the code etc. Also the algorithm doesn't have a name yet, so suggestions are welcome.

**Reply by devemux86** [reply]
@Sublimis Thanks for the screenshots, it seems very nice! :+1: 

- We can certainly merge the improvements in the library.
You can send a pull request when you are ready.

- Are they just improvements in the existing 2 algorithms,
so we do not need a new algorithm?

- Have you checked the performance of the changes?

- Why the screenshots show the Google logo while displaying Mapsforge maps with OpenStreetMap data?

**Reply by Sublimis** [reply]
Screenshots are just for illustrative purposes, the Google logo indicates that Google library was used for making tile screenshots.

The `SimpleShadingAlgorithm` algorithm was a starting point, but it is in fact a completely new algorithm, in its own java file, extending `AbsShadingAlgorithmDefaults`.

The performance is about the same, no notable difference when used. Instead of `sin(sin(sin()))` it computes one `abs()` and one `sqrt()`. Sqrt is just our choice for non-linearity, it could as well be a linear (or any other) function, but this is not the bottleneck.

The real bottleneck for all these algorithms is reading DEM files into a memory, which is ~25MB per file for a 1" resolution.
We noticed that the library maps DEM files from file system to memory, only to later read them sequentially byte by byte (short by short to be exact). This feels like a waste of resources, and can be implemented more efficiently using a simple `BufferedInputStream`. The `DemFileFs.openInputStream()` creating a `BufferedInputStream` is already there. Our algorithm also makes use of this, but it required some changes to the aforementioned files not directly related to the algorithm. You can see some details here: https://github.com/mapsforge/mapsforge/compare/master...Sublimis:mapsforge:master. How do you feel about such a change, abandoning memory mapped files and using a simple stream?

**Reply by devemux86** [reply]
@Sublimis There is no problem to add a new algorithm in the library.

> abandoning memory mapped files and using a simple stream?

Nice, it's better to keep fewer things in memory. I suppose this works on Android and Desktop?

Can you create a pull request, so we can check the new algorithm?

**Reply by Sublimis** [reply]
> Nice, it's better to keep fewer things in memory. I suppose this works on Android and Desktop?

InputStream-s are standard java.io classes, so yes.

> Can you create a pull request, so we can check the new algorithm?

Done!

**Reply by devemux86** [reply, marked as answer]
@Sublimis thanks for the pull request! :+1: 

I merged it in Mapsforge.

**Reply by Sublimis** [reply]
Thank you, it's great to hear this!

**Comment by devemux86** [comment]
@Sublimis 

Which `hgt` files did you use for your screenshots [above](https://github.com/mapsforge/mapsforge/discussions/1512#discussioncomment-10347296)?

**Reply by Sublimis** [reply]
Those are 1 arcsecond files downloaded from the [Viewfinder Panoramas 1" interactive coverage map](https://viewfinderpanoramas.org/Coverage%20map%20viewfinderpanoramas_org1.htm)
